ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,160 @@
+    this.crossColoTracker = crossColoTracker;
+    this.pastDueCounter = pastDueCounter;
+    this.quantile = quantile;
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    super.onResponse(replicaId, isSuccessFul);
+    long elapsedTime;
+    if (unexpiredRequestSendTimes.containsKey(replicaId)) {
+      elapsedTime = time.milliseconds() - unexpiredRequestSendTimes.remove(replicaId);
+    } else {
+      elapsedTime = time.milliseconds() - expiredRequestSendTimes.remove(replicaId);
+    }
+    getTracker(replicaId).update(elapsedTime);
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();
+    return otIterator;
+  }
+
+  /**
+   * Gets the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that {@code replicaId}
+   * belongs to.
+   * @param replicaId the {@link ReplicaId} whose request latency is going to be tracked.
+   * @return the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that
+   * {@code replicaId} belongs to.
+   */
+  private Histogram getTracker(ReplicaId replicaId) {
+    if (replicaId.getDataNodeId().getDatacenterName().equals(datacenterName)) {
+      return localColoTracker;
+    }
+    return crossColoTracker;
+  }
+
+  /**
+   * An iterator to fetch replicas to send requests to. Respects parallelism and discounts requests that have been
+   * outstanding for a certain amount of time from parallelism.
+   */
+  private class OpTrackerIterator implements Iterator<ReplicaId> {
+
+    @Override
+    public boolean hasNext() {
+      return replicaIterator.hasNext() && (inflightCount < parallelism || isOldestRequestPastDue());
+    }
+
+    @Override
+    public void remove() {
+      replicaIterator.remove();
+      if (inflightCount >= parallelism) {
+        // we are here because oldest request is past due
+        Map.Entry<ReplicaId, Long> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
+        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue());
+        unexpiredRequestSendTimes.remove(oldestEntry.getKey());
+        pastDueCounter.inc();
+      }
+      unexpiredRequestSendTimes.put(lastReturned, time.milliseconds());
+      inflightCount++;
+    }
+
+    @Override
+    public ReplicaId next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      lastReturned = replicaIterator.next();
+      return lastReturned;
+    }
+
+    /**
+     * @return {@code true} if the oldest request that was sent has been outstanding for more than the cutoff latency.
+     * @throws IllegalStateException if no requests have been sent yet.
+     */
+    private boolean isOldestRequestPastDue() {
+      if (unexpiredRequestSendTimes.size() == 0) {
+        throw new IllegalStateException("There are no unexpired requests");
```

Comment:
Have added ranges in `RouterConfig` and a check for parallelism >0 in the constructor of `SimpleOperationTracker` (which `AdaptiveOperationTracker` extends from) and gotten rid of this check.